Incompatible with RHEL 9 DEFAULT cryptographic policies

<!--
Thanks for reporting issues back to Nextcloud!

Note: This is the **issue tracker of Nextcloud**, please do NOT use this to get answers to your questions or get help for fixing your installation. This is a place to report bugs to developers, after your server has been debugged. You can find help debugging your system on our home user forums: https://help.nextcloud.com or, if you use Nextcloud in a large organization, ask our engineers on https://portal.nextcloud.com. See also  https://nextcloud.com/support for support options.

Nextcloud is an open source project backed by Nextcloud GmbH. Most of our volunteers are home users and thus primarily care about issues that affect home users. Our paid engineers prioritize issues of our customers. If you are neither a home user nor a customer, consider paying somebody to fix your issue, do it yourself or become a customer.

Guidelines for submitting issues:

* Please search the existing issues first, it's likely that your issue was already reported or even fixed.
    - Go to https://github.com/nextcloud and type any word in the top search/command bar. You probably see something like "We couldn’t find any repositories matching ..." then click "Issues" in the left navigation.
    - You can also filter by appending e. g. "state:open" to the search string.
    - More info on search syntax within github: https://help.github.com/articles/searching-issues
    
* This repository https://github.com/nextcloud/end_to_end_encryption/issues is *only* for issues within the Nextcloud End-to-End-Encryption Server app. Please report client issues related to End-to-End-Encryption in the corresponding client repository:
    - Android: http://github.com/nextcloud/android/issues
    - iOS / iPadOS: http://github.com/nextcloud/ios/issues
    - Linux / macOS / Windows: https://github.com/nextcloud/desktop/issues
  
* SECURITY: Report any potential security bug to us via our HackerOne page (https://hackerone.com/nextcloud) following our security policy (https://nextcloud.com/security/) instead of filing an issue in our bug tracker.  

* The issues in other components should be reported in their respective repositories: You will find them in our GitHub Organization (https://github.com/nextcloud/)
  
* You can also use the Issue Template app to prefill most of the required information: https://apps.nextcloud.com/apps/issuetemplate
-->

<!--- Please keep this note for other contributors -->

### How to use GitHub

* Please use the 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to show that you are affected by the same issue.
* Please don't comment if you have no relevant information to add. It's just extra noise for everyone subscribed to this issue.


### Steps to reproduce
1. Create an encrypted folder via the web browser
2. 
3.

### Expected behaviour
A mnemonic is generated and the encrypted folder is created.

### Actual behaviour
No mnemonic is generated and the encrypted folder is not created. The following error is included in a notification: Creating new encrypted folder: Failed.

### Server configuration

**Operating system:**
Rocky Linux 9
**Web server:**
Apache
**Database:**
PostgreSQL 16
**PHP version:**
8.3
**Nextcloud version:** (see Nextcloud admin page)
33.0.1.2
**Updated from an older Nextcloud/ownCloud or fresh install:**
Updated
**Where did you install Nextcloud from:**
https://download.nextcloud.com/server/releases/
**Signing status:**
<details>
<summary>Signing status</summary>

```
No errors have been found.
```
</details>

**List of activated apps:**
<details>
<summary>App list</summary>

```
Enabled:
  - activity: 6.0.0-dev.0
  - bruteforcesettings: 6.0.0-dev.0
  - circles: 33.0.0
  - cloud_federation_api: 1.17.0
  - comments: 1.23.0
  - contacts: 8.4.2
  - contactsinteraction: 1.14.1
  - dashboard: 7.13.0
  - dav: 1.36.0
  - deck: 1.17.0
  - end_to_end_encryption: 2.0.0
  - federatedfilesharing: 1.23.0
  - federation: 1.23.0
  - files: 2.5.0
  - files_downloadlimit: 5.1.0-dev.0
  - files_pdfviewer: 6.0.0-dev.0
  - files_reminders: 1.6.0
  - files_sharing: 1.25.2
  - files_trashbin: 1.23.0
  - files_versions: 1.26.0
  - firstrunwizard: 6.0.0-dev.0
  - groupfolders: 21.0.6
  - guests: 4.6.0
  - logreader: 6.0.0
  - lookup_server_connector: 1.21.0
  - nextcloud_announcements: 5.0.0
  - notifications: 6.0.0
  - oauth2: 1.21.0
  - password_policy: 5.0.0-dev.0
  - photos: 6.0.0-dev.0
  - privacy: 5.0.0-dev.0
  - profile: 1.2.0
  - provisioning_api: 1.23.0
  - recommendations: 6.0.0-dev.0
  - related_resources: 4.0.0-dev.0
  - richdocuments: 10.1.2
  - serverinfo: 5.0.0-dev.0
  - settings: 1.16.0
  - sharebymail: 1.23.0
  - support: 5.0.0
  - survey_client: 5.0.0-dev.0
  - systemtags: 1.23.0
  - text: 7.0.0-dev.3
  - theming: 2.8.0
  - twofactor_backupcodes: 1.22.0
  - twofactor_totp: 15.0.0-dev.0
  - updatenotification: 1.23.0
  - user_oidc: 8.7.0
  - user_status: 1.13.0
  - viewer: 6.0.0-dev.0
  - weather_status: 1.13.0
  - webhook_listeners: 1.5.0
  - workflowengine: 2.15.0
Disabled:
  - admin_audit: 1.23.0
  - app_api: 33.0.0 (installed 32.0.0)
  - encryption: 2.21.0
  - files_external: 1.25.1
  - suspicious_login: 11.0.0-dev.0
  - testing: 1.23.0
  - twofactor_nextcloud_notification: 7.0.0
  - user_ldap: 1.24.0
```
</details>

**Nextcloud configuration:**
<details>
<summary>Config report</summary>

```
{
    "system": {
        "trusted_domains": [
            "..."
        ],
        "redis": {
            "host": "***REMOVED SENSITIVE VALUE***",
            "port": 0,
            "timeout": 1.5
        },
        "memcache.local": "\\OC\\Memcache\\APCu",
        "memcache.locking": "\\OC\\Memcache\\Redis",
        "memcache.distributed": "\\OC\\Memcache\\Redis",
        "overwrite.cli.url": "...",
        "overwriteprotocol": "https",
        "htaccess.RewriteBase": "\/",
        "maintenance_window_start": 7,
        "passwordsalt": "***REMOVED SENSITIVE VALUE***",
        "secret": "***REMOVED SENSITIVE VALUE***",
        "datadirectory": "***REMOVED SENSITIVE VALUE***",
        "dbtype": "pgsql",
        "version": "33.0.1.2",
        "dbname": "***REMOVED SENSITIVE VALUE***",
        "dbhost": "***REMOVED SENSITIVE VALUE***",
        "dbtableprefix": "oc_",
        "dbuser": "***REMOVED SENSITIVE VALUE***",
        "dbpassword": "***REMOVED SENSITIVE VALUE***",
        "installed": true,
        "instanceid": "***REMOVED SENSITIVE VALUE***",
        "maintenance": false,
        "loglevel": 2
    }
}
```
</details>

**Are you using external storage, if yes which one:** local/smb/sftp/...
No
**Are you using encryption:** yes/no
No
**Are you using an external user-backend, if yes which one:** LDAP/ActiveDirectory/Webdav/...
No
### Client configuration
**Browser:**
Firefox
**Operating system:**
Windows 11
### Logs

<!--- Reports without logs might be closed as unqualified reports! -->

#### Web server error log
<details>
<summary>Web server error log</summary>

```
Insert your webserver log here
```
</details>

#### Nextcloud log (data/nextcloud.log)
<details>
<summary>Nextcloud log</summary>

```
[end_to_end_encryption] Error: could not sign the CSR, please make sure to submit a valid CSR
	POST /ocs/v2.php/apps/end_to_end_encryption/api/v2/public-key
	from xxx by xxx at Mar 26, 2026, 4:34:06 PM

[PHP] Warning: openssl_csr_sign(): Failed to sign it at /var/www/nextcloud/apps/end_to_end_encryption/lib/SignatureHandler.php#38
	POST /ocs/v2.php/apps/end_to_end_encryption/api/v2/public-key
	from xxx by xxx at Mar 26, 2026, 4:34:06 PM
```
</details>

#### Browser log
<details>
<summary>Browser log</summary>

```
XHRPOST
https://domain.ca/ocs/v2.php/apps/end_to_end_encryption/api/v2/public-key
[HTTP/2 400  53ms]

[ERROR] end_to_end_encryption: Error creating public key 
Object { app: "end_to_end_encryption", uid: "xxx", level: 2, error: AxiosError }
_plugin-vue_export-helper-DUspQHA4.chunk.mjs:4:381
    log index.mjs:47
    error index.mjs:65
    Kk api.ts:148
    SN encryptionService.ts:42
    n CreateFolderDialogStepSetupEncryption.vue:38
    o CreateFolderDialog.vue:70
    Ci runtime-core.esm-bundler.js:199
    Wt runtime-core.esm-bundler.js:206
    n3 runtime-core.esm-bundler.js:4380
    h NcButton-CzpKEx4V.mjs:84
    Ci runtime-core.esm-bundler.js:199
    Wt runtime-core.esm-bundler.js:206
    u runtime-dom.esm-bundler.js:731
    (Async: EventListener.handleEvent)
    W3 runtime-dom.esm-bundler.js:681
    K3 runtime-dom.esm-bundler.js:699
    ef runtime-dom.esm-bundler.js:776
    V runtime-core.esm-bundler.js:5596
    U runtime-core.esm-bundler.js:5534
    h runtime-core.esm-bundler.js:5391
    G runtime-core.esm-bundler.js:6052
    run reactivity.esm-bundler.js:237
    W runtime-core.esm-bundler.js:6179
    M runtime-core.esm-bundler.js:5954
    _ runtime-core.esm-bundler.js:5905
    h runtime-core.esm-bundler.js:5403
    P runtime-core.esm-bundler.js:5656
    x runtime-core.esm-bundler.js:5835
    h runtime-core.esm-bundler.js:5378
    P runtime-core.esm-bundler.js:5656
    V runtime-core.esm-bundler.js:5578
    U runtime-core.esm-bundler.js:5534
    h runtime-core.esm-bundler.js:5391
    P runtime-core.esm-bundler.js:5656
    V runtime-core.esm-bundler.js:5578
    U runtime-core.esm-bundler.js:5534
    h runtime-core.esm-bundler.js:5391
    P runtime-core.esm-bundler.js:5656
    x runtime-core.esm-bundler.js:5835
    h runtime-core.esm-bundler.js:5378
    P runtime-core.esm-bundler.js:5656
    V runtime-core.esm-bundler.js:5578
    U runtime-core.esm-bundler.js:5534
    h runtime-core.esm-bundler.js:5391
[ERROR] end_to_end_encryption: Error while creating new encrypted folder: 
Object { app: "end_to_end_encryption", uid: "xxx", level: 2, error: AxiosError }
_plugin-vue_export-helper-DUspQHA4.chunk.mjs:4:381
    log index.mjs:47
    error index.mjs:65
    handler new-encrypted-folder.ts:46
    onClick index-DPkUyJ8g.mjs:1473
    click index-DPkUyJ8g.mjs:1581
    VueJS 4
    onClick actionText-BMig9Egt.mjs:78
    handleClick NcActionButton-1gSJfFUC.mjs:145
    VueJS 33
```
</details>

The app appears to be using SHA-1. I'm able to mitigate the issue by switching the system-level cryptographic policy to [DEFAULT:SHA1](https://docs.redhat.com/en/documentation/red_hat_enterprise_linux/9/html/security_hardening/using-the-system-wide-cryptographic-policies_security-hardening#proc_re-enabling-sha-1_using-the-system-wide-cryptographic-policies), but preferably the app wouldn't be using that algorithm at any point.

EDIT:
I took a closer look at the PHP documentation for the offending function `openssl_csr_sign`, and for some reason it's defaulting to SHA-1 in my setup rather than the `default_md` in my openssl.cnf, which is SHA-256, so I think some other issue it at play here in RHEL 9/PHP. I tested adding `array('digest_alg'=>'sha256')` to `$options` in `openssl_csr_sign`, and everything seems to be working as it should. I'll try to throw together a PR if I can find some time.

--- a/lib/SignatureHandler.php
+++ b/lib/SignatureHandler.php
@@ -34,7 +34,13 @@ class SignatureHandler {
 	 */
 	public function sign(string $csr): string {
 		$systemKeys = $this->identityProofManager->getSystemKey();
-		$signedCertificate = openssl_csr_sign($csr, null, $systemKeys->getPrivate(), $this->validity);
+		$signedCertificate = openssl_csr_sign(
+			$csr,
+			null,
+			$systemKeys->getPrivate(),
+			$this->validity,
+			['digest_alg' => 'sha256'],
+		);
 		if ($signedCertificate === false) {
 			throw new BadMethodCallException('could not sign the CSR, please make sure to submit a valid CSR');
 		}

